core/variable_operations.cpp

```diff
@@ -294,4 +294,88 @@ Variable masks_merge_if_contained(const MasksConstProxy &masks,
   return mask_union;
 }
 
+namespace {
+void validate_inputs_for_replacement(const VariableConstProxy &var,
+                                     const VariableConstProxy &replacement) {
+  if (var.dtype() != replacement.dtype())
+    throw except::TypeError("Replacement type doesn't match type of input");
+  if (replacement.data().size() != 1)
+    throw except::SizeError("Single value expected for replacement");
+}
+} // namespace
+
+VariableProxy nan_to_num(const VariableConstProxy &var,
+                         const VariableConstProxy &replacement,
+                         const VariableProxy &out) {
+
+  validate_inputs_for_replacement(var, replacement);
+
+  transform_in_place<pair_self_t<double, float>>(
+      out, var,
+      scipp::overloaded{
+          [&](auto &out_v, const auto &in_v) {
+            using std::isnan;
+            using V_OUT = std::decay_t<decltype(out_v)>;
+            using V_IN = std::decay_t<decltype(in_v)>;
+            if constexpr (is_ValueAndVariance_v<V_OUT>) {
+              if constexpr (is_ValueAndVariance_v<V_IN>) {
+                using value_type = typename V_IN::value_type;
+                if (isnan(in_v)) {
+                  out_v.value = replacement.value<value_type>();
+                  out_v.variance = replacement.hasVariances()
+                                       ? replacement.variance<value_type>()
+                                       : out_v.value;
+                } else {
+                  // Assign directly to output
+                  out_v.value = in_v.value;
+                  out_v.variance = in_v.variance;
+                }
+              } else {
+                if (isnan(in_v)) {
+                  out_v.value = replacement.value<V_IN>();
+                  out_v.variance = replacement.hasVariances()
+                                       ? replacement.variance<V_IN>()
+                                       : out_v.value;
+                } else {
+                  // Assign directly to output
+                  out_v.value = in_v;
+                  out_v.variance = in_v;
+                }
+              }
+            } else {
+              out_v = isnan(in_v) ? replacement.value<V_IN>() : in_v;
+            }
+          },
+          [&](units::Unit &ua, const units::Unit &ub) {
+            expect::equals(ua, ub);
+          }});
+  return out;
+}
```

Comment:
If this is the shortest way to write this function I would argue to remove it until we have a performance benchmark for a real-world application where the operation with out arg is crucial. Relative to its importance this is way to long and complex to be maintainable.
As a compromise we may want to support in-place operation where `out` is the input. That should cover most performance-relevant cases.